src/storage/mvcc/reader/reader.rs

```diff
@@ -278,12 +279,103 @@ impl<S: EngineSnapshot> MvccReader<S> {
         Ok(())
     }
 
-    pub fn load_in_memory_pessimisitic_lock_range<F>(
+    /// Scan all types of locks(pessimitic, prewrite) satisfying `filter`
+    /// condition from both in-memory pessimitic lock table and the storage
+    /// within [start_key, end_key) .
+    pub fn scan_locks<F>(
+        &mut self,
+        start_key: Option<&Key>,
+        end_key: Option<&Key>,
+        filter: F,
+        limit: usize,
+        source: ScanLockReadTimeSource,
+    ) -> Result<(Vec<(Key, Lock)>, bool)>
+    where
+        F: Fn(&Key, TxnLockRef<'_>) -> bool,
+    {
+        let (memory_locks, memory_has_remain) = self.load_in_memory_pessimistic_lock_range(
+            start_key,
+            end_key,
```

Comment:
Reminded by https://github.com/tikv/tikv/pull/16180#discussion_r1430442189 , should we limit the scan range here to `[max(start_key, snapshot.lower_bound), min(end_key, snapshot.upper_bound))`? For example, let
```
snap1: { lower: k0, upper: k1, locks: [l0] }
snap2: { lower: k1, upper: k2, locks: [l1] }
...
mem: [
    l2, // k0 < l2.key < k1
    l3, // k2 < l3.key
    ...
]
```
would `scan_locks(limit=3)` return `[l0,l2,l3]`? (note `l2.key < l1.key < l3.key`)

Reply:
The `scan_lock` is not a snapshot read when considering in-memory pessimistic locks. The storage read and memory read could be processed independently to get locks within [start_key, end_key) I think.

I don't quite undestand the example above, could you explain more about it?

Reply:
Let:
```
snap1: { lower: nil, upper: 'b', locks: [l1{key:'a1'}] }
snap2: { lower: 'b', upper: 'c', locks: [l2{key:'b1'}] }
snap3: { lower: 'c', upper: nil, locks: [l3{key:'c1'}] }
mem: [
    ml1{key:'a2'},
    ml2{key:'c2'},
]
```
According to the following gc resolve lock logic:
```go
ResolveLocksForRange(startKey, endKey) {
    key := startKey
    for {
        locks, loc, err := resolver.ScanLocksInOneRegion(bo, key, maxVersion, scanLimit)
        if len(locks) < int(scanLimit) {
            stat.CompletedRegions++
            key = loc.EndKey
        } else {
            key = locks[len(locks)-1].Key
        }
        if len(key) == 0 || (len(endKey) != 0 && bytes.Compare(key, endKey) >= 0) {
            break
        }
    }
}
```
l2 and l3 might be omitted? since
- ResolveLocksForRange(nil, nil, scanLimit=3)
  - ScanLocksInOneRegion(nil, 3) -> [l1, ml1, ml2]
  - ScanLocksInOneRegion('c2', 3) -> []

Reply:
The above case would not happen because the end key of region is specified [here](https://github.com/tikv/client-go/blob/c7d29aafa7073f64555532d0d10f19714bc8e6c2/tikv/gc.go#L229C13-L229C13). I just worry about if there are some similar issues.

Reply:
Oh, my fault, I forget that the mem keys are also managed by each region.

Reply:
I see, thanks for the reminding.
The problem is that the next scan key could not be the `lock[len(locks - 1)].key` after introducing  the `merge` operation for memory and storage locks.

For the `scan_lock` itself, it would be easier to use just considering [start_key, end_key), another approach to resolve this is to return `next_scan_key` explicitly in the `scan_lock` responses if there are remaining results, and the `next_scan_key` could be `min(snapshot_start_key, lock[len(locks - 1)].key)`.

Reply:
@zyguan @MyonKeminta 
What do you think? Any other ideas?

Reply:
It seems that unlike the `Scanner`which would do `verify_range` there is no region range check when scanning locks from storage by now.
@MyonKeminta @zyguan 
Is it necessary to add such checks for scanning locks?

Reply:
I think it's probably unnecessary, [here](https://github.com/tikv/tikv/blob/master/components/raftstore/src/store/txn_ext.rs#L29) mem locks are guaranteed to be within the range of the region?

Reply:
It might be a good idea to return a KeyNotInRegion error if the specified start_key or end_key is non-empty and out of the range of the region. TiDB's GCWorker is expected to get a region info and limit the range specified in the request within the region's range. Then, if TiKV here gets a snapshot successfully without reporting an EpochNotMatch error but the specified range exceeds the region, it must be caused by TiDB sending request wrongly.
But it still doesn't make full sense, as an empty start_key or end_key, which represents unlimited, is in fact still limited by the region range.
So I think such a check might make sense, but it's not strongly necessary. Of course it's not necessary to be done with in this PR, after all